cmd/sink/expr/cel.go

```diff
@@ -0,0 +1,211 @@
+	archive map[string]cel.Program
+	delete  map[string]cel.Program
+}
+
+// Returns a Filters struct with an empty archive and delete slice
+func EmptyFilters() *Filters {
+	return &Filters{
+		archive: make(map[string]cel.Program),
+		delete:  make(map[string]cel.Program),
+	}
+}
+
+// Creates a Filters struct from the path to a directory where a ConfigMap was mounted. If path is empty string or path
+// does not exist, it returns a Filters struct with empty archive and delete slices. It will attempt to create all the
+// cel programs that it can from the ConfigMap. Any errors that are encountered are wrapped together and returned. Even
+// if this function returns an error, the Filters struct returned can still be used and will not be nil.
+func NewFilters(path string) (*Filters, error) {
+	var errList []error
+	filters := &Filters{
+		archive: make(map[string]cel.Program),
+		delete:  make(map[string]cel.Program),
+	}
+	exists, err := files.PathExists(path)
+	if err != nil {
+		return filters, fmt.Errorf("cannot determine if ConfigMap is mounted at path %s: %s", path, err)
+	}
+	if path == "" || !exists {
+		return filters, fmt.Errorf("cannot create Filters. ConfigMap is not mounted at path %s", path)
+	}
+	filterFiles, err := files.DirectoryFiles(path)
+	if err != nil {
+		return filters, fmt.Errorf(
+			"cannot create Filters. Could not read files created from ConfigMap mounted at path %s: %s",
+			path,
+			err,
+		)
+	}
+	for namespace, filePath := range filterFiles {
+		resourcesBytes, err := os.ReadFile(filePath)
+		if err != nil {
+			errList = append(errList, err)
+			continue
+		}
+		resources := []v1alpha1.KubeArchiveConfigResource{}
+		err = yaml.Unmarshal(resourcesBytes, &resources)
+		if err != nil {
+			errList = append(errList, err)
+			continue
+		}
+		for _, resource := range resources {
+			namespaceGvk := NamespaceGroupVersionKind(namespace, resource)
+			archiveExpr, err := CreateCelExpr(resource.ArchiveWhen)
+			if err != nil {
+				errList = append(errList, err)
+			} else {
+				filters.archive[namespaceGvk] = *archiveExpr
+			}
+			deleteExpr, err := CreateCelExpr(resource.DeleteWhen)
+			if err != nil {
+				errList = append(errList, err)
+			} else {
+				filters.delete[namespaceGvk] = *deleteExpr
+			}
+		}
+	}
+
+	return filters, errors.Join(errList...)
+
+}
+
+// Returns whether obj needs to be archived. Obj needs to be archived if any of the cel programs in archive return true
+// or if obj needs to be deleted. If obj is nil, it returns false
+func (f *Filters) MustArchive(ctx context.Context, obj *unstructured.Unstructured) bool {
+	if obj == nil {
+		return false
+	}
+	mustArchive := f.MustDelete(ctx, obj)
+	if mustArchive {
+		return mustArchive
+	}
+	globalGvk := GlobalGroupVersionKind(obj)
+	namespaceGvk := NamespaceGroupVersionKindFromObject(obj)
+	program, exists := f.archive[globalGvk]
+	if exists {
+		mustArchive = executeCel(ctx, program, obj)
+	}
+	if mustArchive {
+		return mustArchive
+	}
```

Comment:
Since the second conditional is only relevant if the first conditional is true, it should be in the block:

```suggestion
	if exists {
		mustArchive = executeCel(ctx, program, obj)
	    if mustArchive {
		    return mustArchive
	    }
	}
```

Reply:
This would all be much cleaner if you only evaluate one expression for both archive and delete. What needs to happen is that the `global` expressions for archive and delete need to be `or`'d to each of the namespace expressions. Then evaluate those.

Reply:
I changed the logic for creating filters which allowed me to remove this logic